core/ecschema-metadata/src/Metadata/Class.ts

```diff
@@ -612,76 +633,54 @@ export abstract class ECClass extends SchemaItem implements CustomAttributeConta
-   * @param overwriteExisting
-   *
-   * @internal
-   */
-  protected static mergeProperties(target: Property[], existingValues: Map<string, number>, propertiesToMerge: Iterable<Property>, overwriteExisting: boolean) {
-    for (const property of propertiesToMerge) {
-      const upperCaseName = property.name.toUpperCase();
-      const existing = existingValues.get(upperCaseName);
-      if (existing !== undefined) {
-        if (overwriteExisting) {
-          target[existing] = property;
-        }
-      } else {
-        existingValues.set(upperCaseName, target.length);
-        target.push(property);
-      }
-    }
-  }
-
-  /**
-   *
-   * @param result
-   * @param existingValues
+   * @param cache
    * @returns
    *
    * @internal
    */
-  protected async buildPropertyCache(result: Property[], existingValues?: Map<string, number>): Promise<void> {
-    if (!existingValues) {
-      existingValues = new Map<string, number>();
-    }
-
-    if (this.baseClass) {
-      const baseClass = await this.baseClass;
-      if (baseClass) {
-        ECClass.mergeProperties(result, existingValues, await baseClass.getProperties(), false);
+protected async buildPropertyCache(): Promise<Map<string, Property>> {
+  const cache = new Map<string, Property>();
+  const baseClass = await this.baseClass;
+  if (baseClass) {
+    Array.from(baseClass.getPropertiesSync()).forEach(property => {
+      if (!cache.has(property.name.toUpperCase())) {
+        cache.set(property.name.toUpperCase(), property);
       }
-    }
-
-    if (!this._properties)
-      return;
+    });
+  }
 
-    ECClass.mergeProperties(result, existingValues, [...this._properties.values()], true);
+  if (this._properties) {
+    this._properties.forEach(property => {
+      cache.set(property.name.toUpperCase(), property);
+    });
   }
+  return cache;
+}
 
   /**
    *
-   * @param result
-   * @param existingValues
+   * @param cache
    * @returns
    *
    * @internal
    */
-  protected buildPropertyCacheSync(result: Property[], existingValues?: Map<string, number>): void {
-    if (!existingValues) {
-      existingValues = new Map<string, number>();
-    }
-
+  protected buildPropertyCacheSync(): Map<string, Property> {
+    const cache = new Map<string, Property>();
     const baseClass = this.getBaseClassSync();
     if (baseClass) {
-      ECClass.mergeProperties(result, existingValues, baseClass.getPropertiesSync(), false);
+      Array.from(baseClass.getPropertiesSync()).forEach(property => {
+        if (!cache.has(property.name.toUpperCase())) {
+          cache.set(property.name.toUpperCase(), property);
+        }
+      });
     }
 
-    if (!this._properties)
-      return;
-
-    ECClass.mergeProperties(result, existingValues, [...this._properties.values()], true);
+    if (this._properties) {
+      this._properties.forEach(property => {
+        cache.set(property.name.toUpperCase(), property);
+      });
+    }
+    return cache;
   }
```

Comment:
Why do we have two methods here, doing the same thing? The only difference is that the base class is awaited in the in the async method and in the sync, we use the sync lookup method. 

I guess, we could use the sync method for both cases, or we pass the resolved baseclass into the method (but I'm not sure whether I like that approach)